### DISCUSSION TITLE
How to use GLCD Fonts

### DISCUSSION THREAD

**Discussion by igotab** [Bodmer/TFT_eSPI#1752] [Q&A - Fonts] [answered]
I have a number of old GLCD fonts generated by MicroE GLCD Font Creator (https://www.mikroe.com/glcd-font-creator)
that I would like to use with TFT_eSPI libarary. The fonts are in byte arrays format similar to the included glcdfont.c font.

I can't find any references on how to define the size and character range for my fonts and load them. 

Can someone help  please.

**Comment by Bodmer** [comment, marked as answer]
The library does not support that font format. Formats supported are Free fonts, VLW smooth fonts and some built-in high update speed fonts in a bespoke RLE format.

There are companion libraries that provide access to 1000's of ready built fonts and you may find one of those suitable:
https://github.com/Bodmer/TFT_eSPI_ext
https://github.com/Bodmer/U8g2_for_TFT_eSPI

**Reply by igotab** [reply]
Thank you for your support.

Is there a way to increase the spacing between characters in either free or smooth fonts? 
Both look a bit dense, especially the lower case characters.

**Reply by Bodmer** [reply]
There is no capability built into the library to adjust the character kerning/tracking. It is possible to either modify the font file or write code to do the spacing adjustment within a rendering function. The spacing varies on a per glyph basis for the proportional fonts.

The free font files are quite easy to edit in this respect. The font metrics are [stored in a structure](https://github.com/Bodmer/TFT_eSPI/blob/master/Fonts/GFXFF/gfxfont.h#L13-L20). You need to change xAdvance and/or xOffset. Increasing xAdvance will move the next cursor position right, so that there is a bigger gap to the next character. Increasing xOffset will increase the gap before the chanracter.

Example metrics for "a" [here](https://github.com/Bodmer/TFT_eSPI/blob/master/Fonts/GFXFF/FreeSansBold12pt7b.h#L252). In this case xAdvance is 14 and xOffset is 1, so those are the values to change for that font.

Adjusting the spacing for smooth fonts is also possible but you would need to either modify the VLW file or adust the metrics after the font has been loaded (after loadFont...). In this case the metrics are loaded into RAM and so the position adjustment could be done in the code. Smooth font VLW file format is [detailed here](https://github.com/Bodmer/TFT_eSPI/blob/master/Extensions/Smooth_font.cpp#L39-L94), and how metrics are stored in memory after the font is loaded are [detailed here](https://github.com/Bodmer/TFT_eSPI/blob/master/Extensions/Smooth_font.cpp#L194-L199).

You can create you own free fonts using online wed based tools and VLW format fonts using the Processing sketch provided in the tools folder.

You may prefer to use mono spaced fonts if you are used to LCD style spacings.

**Comment by igotab** [comment]
Thank you for your great support
